docs/docs/tutorials/cache/index.md

```diff
@@ -0,0 +1,185 @@
+# Use and Customize DSPy Cache
+
+In this tutorial, we will explore the design of DSPy's caching mechanism and demonstrate how to effectively use and customize it.
+
+## DSPy Cache Structure
+
+DSPy's caching system is architected in three distinct layers:
+
+1.  **In-memory cache**: Implemented using `cachetools.LRUCache`, this layer provides fast access to frequently used data.
+2.  **On-disk cache**: Leveraging `diskcache.FanoutCache`, this layer offers persistent storage for cached items.
+3.  **Prompt cache (Server-side cache)**: This layer is managed by the LLM service provider (e.g., OpenAI, Anthropic).
+
+While DSPy does not directly control the server-side prompt cache, it offers users the flexibility to enable, disable, and customize the in-memory and on-disk caches to suit their specific requirements.
+
+## Using DSPy Cache
+
+By default, both in-memory and on-disk caching are automatically enabled in DSPy. No specific action is required to start using the cache. When a cache hit occurs, you will observe a significant reduction in the module call's execution time. Furthermore, if usage tracking is enabled, the usage metrics for a cached call will be `None`.
+
+Consider the following example:
+
+```python
+import dspy
+import os
+import time
+
+os.environ["OPENAI_API_KEY"] = "{your_openai_key}"
+
+dspy.settings.configure(lm=dspy.LM("openai/gpt-4o-mini"), track_usage=True)
+
+predict = dspy.Predict("question->answer")
+
+start = time.time()
+result1 = predict(question="Who is the GOAT of basketball?")
+print(f"Time elapse: {time.time() - start: 2f}\n\nTotal usage: {result1.get_lm_usage()}")
+
+start = time.time()
+result2 = predict(question="Who is the GOAT of basketball?")
+print(f"Time elapse: {time.time() - start: 2f}\n\nTotal usage: {result2.get_lm_usage()}")
+```
+
+A sample output looks like:
+
+```
+Time elapse:  4.384113
+Total usage: {'openai/gpt-4o-mini': {'completion_tokens': 97, 'prompt_tokens': 144, 'total_tokens': 241, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0, 'text_tokens': None}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0, 'text_tokens': None, 'image_tokens': None}}}
+
+Time elapse:  0.000529
+Total usage: {}
+```
+
+## Disabling/Enabling DSPy Cache
+
+There are scenarios where you might need to disable caching, either entirely or selectively for in-memory or on-disk caches. For instance:
+
+- You require different responses for identical LM requests.
+- You lack disk write permissions and need to disable the on-disk cache.
+- You have limited memory resources and wish to disable the in-memory cache.
+
+DSPy provides the `dspy.configure_cache()` utility function for this purpose. You can use the corresponding flags to control the enabled/disabled state of each cache type:
```

Comment:
Same for `dspy.configure_cache()`. Wouldn't it be better if we enable users to configure cache setting through `dspy.settings.configure`?